### ISSUE TITLE
Extremely slow TFLite inference on Jetson Xavier NX

### ISSUE THREAD

**Issue by isuchy** [ultralytics/ultralytics#18605]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Predict

### Bug

Hey guys,

I'm doing a benchmark of object detection in video on edge devices. From the Nvidia Jetson product line I have the older Xavier NX.  

I have exported the YOLOv8l model (fine-tuned on custom dataset) to TFLite float32 and also fully quantized to int8. When I run the test script with video (see below), the frame inference results have terribly high times, over 20000 ms!   

Tensorflow Lite doesn't support GPU on aarch64, so inferencing only runs on CPU, with XNNPACK delegate.
TFLite is of course not a recommended format for Jetson boards, it's just a benchmark, but I wonder if it's possible to do something else with it to improve performance.

Maybe I'm just doing something wrong that I'm missing so far, so any ideas? :)

```
Processing video with inference...
Original video FPS: 29.97
Loading best_integer_quant.tflite for TensorFlow Lite inference...
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
WARNING ⚠️ Metadata not found for 'model=best_integer_quant.tflite'

0: 640x640 3 class0s, 24442.3ms
Speed: 35.6ms preprocess, 24442.3ms inference, 30.3ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 4 class0s, 21152.1ms
Speed: 8.8ms preprocess, 21152.1ms inference, 4.0ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 2 class0s, 21024.6ms
Speed: 13.6ms preprocess, 21024.6ms inference, 4.0ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 1 class0, 21973.4ms
Speed: 12.7ms preprocess, 21973.4ms inference, 4.5ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 1 class0, 19933.4ms
Speed: 14.8ms preprocess, 19933.4ms inference, 3.8ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 3 class0s, 22208.8ms
Speed: 11.1ms preprocess, 22208.8ms inference, 3.9ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 3 class0s, 22295.3ms
Speed: 15.1ms preprocess, 22295.3ms inference, 4.1ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 3 class0s, 22030.5ms
Speed: 12.6ms preprocess, 22030.5ms inference, 4.0ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 5 class0s, 21100.4ms
Speed: 16.5ms preprocess, 21100.4ms inference, 4.0ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 3 class0s, 22286.7ms
Speed: 12.7ms preprocess, 22286.7ms inference, 3.8ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 1 class0, 22117.8ms
Speed: 13.5ms preprocess, 22117.8ms inference, 4.2ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 3 class0s, 19420.4ms
Speed: 12.9ms preprocess, 19420.4ms inference, 3.9ms postprocess per image at shape (1, 3, 640, 640)
...

```

With the TFLite float32 model, inference is approximately 50% faster, but those times are still extremely high.
```
Loading best_float32.tflite for TensorFlow Lite inference...
INFO: Created TensorFlow Lite XNNPACK delegate for CPU.
WARNING ⚠️ Metadata not found for 'model=best_float32.tflite'

0: 640x640 6 class0s, 12531.8ms
Speed: 24.2ms preprocess, 12531.8ms inference, 8.9ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 8 class0s, 12326.6ms
Speed: 21.9ms preprocess, 12326.6ms inference, 3.7ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 class0s, 12408.4ms
Speed: 7.7ms preprocess, 12408.4ms inference, 3.6ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 class0s, 12428.9ms
Speed: 21.1ms preprocess, 12428.9ms inference, 3.6ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 class0s, 12427.9ms
Speed: 21.3ms preprocess, 12427.9ms inference, 3.6ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 4 class0s, 12436.1ms
Speed: 17.4ms preprocess, 12436.1ms inference, 3.5ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 5 class0s, 12418.2ms
Speed: 17.4ms preprocess, 12418.2ms inference, 3.8ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 5 class0s, 12452.3ms
Speed: 15.1ms preprocess, 12452.3ms inference, 3.8ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 class0s, 12422.7ms
Speed: 17.8ms preprocess, 12422.7ms inference, 3.6ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 6 class0s, 12435.4ms
Speed: 17.1ms preprocess, 12435.4ms inference, 3.7ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 5 class0s, 12433.9ms
Speed: 15.5ms preprocess, 12433.9ms inference, 3.5ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 4 class0s, 12471.3ms
Speed: 17.3ms preprocess, 12471.3ms inference, 3.7ms postprocess per image at shape (1, 3, 640, 640)

0: 640x640 4 class0s, 12421.7ms
Speed: 18.3ms preprocess, 12421.7ms inference, 3.6ms postprocess per image at shape (1, 3, 640, 640)
...
```

### Environment

Ultralytics 8.3.58 🚀 Python-3.8.10 torch-2.1.0a0+41361538.nv23.06 CUDA:0 (Xavier, 7513MiB)
Setup complete ✅ (6 CPUs, 7.3 GB RAM, 41.1/118.4 GB disk)

OS                  Linux-5.10.104-tegra-aarch64-with-glibc2.29
Environment         Linux
Python              3.8.10
Install             pip
RAM                 7.34 GB
Disk                41.1/118.4 GB
CPU                 ARMv8 Processor rev 0 (v8l)
CPU count           6
GPU                 Xavier, 7513MiB
GPU count           1
CUDA                11.4

numpy               ✅ 1.23.5>=1.23.0
numpy               ✅ 1.23.5<2.0.0; sys_platform == "darwin"
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ✅ 4.10.0.84>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 5.3.1>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.10.1>=1.4.1
torch               ✅ 2.1.0a0+41361538.nv23.6>=1.8.0
torch               ✅ 2.1.0a0+41361538.nv23.6!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.16.1+fdea156>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0

### Minimal Reproducible Example

import cv2
import time
from ultralytics import YOLO

# Load the local YOLO v8 model
model = YOLO('best_integer_quant.tflite') 

def calculate_average_fps(frame_count, total_time):
    return frame_count / total_time if total_time > 0 else 0

def process_video(video_path, with_inference=True, respect_playback_rate=True):
    # Open the video file using OpenCV
    cap = cv2.VideoCapture(video_path)
    if not cap.isOpened():
        print(f"Error: Cannot open video {video_path}")
        return

    # Get the video's original FPS from metadata
    original_fps = cap.get(cv2.CAP_PROP_FPS)
    frame_duration = 1.0 / original_fps if original_fps > 0 else 0
    print(f"Original video FPS: {original_fps:.2f}")

    frame_count = 0
    total_inference_time = 0.0
    start_time = time.time()

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        frame_start_time = time.time()
        frame_count += 1

        if with_inference:
            # Perform inference directly with the YOLO model
            inference_start = time.time()
            results = model.predict(frame, imgsz=640, conf=0.25)
            inference_end = time.time()
            
            frame_inference_time = inference_end - inference_start
            total_inference_time += frame_inference_time

        # Respect the playback frame duration
        if respect_playback_rate and frame_duration > 0:
            elapsed = time.time() - frame_start_time
            remaining_time = frame_duration - elapsed
            if remaining_time > 0:
                time.sleep(remaining_time)

    end_time = time.time()
    total_time = end_time - start_time

    # Calculate average FPS and inference time
    avg_fps = calculate_average_fps(frame_count, total_time)
    avg_inference_time = total_inference_time / frame_count if frame_count > 0 else 0
    cap.release()

    return avg_fps, avg_inference_time

def main():
    video_path = "sample.mp4"  # Replace with your video file path
    print("Processing video with inference...")
    avg_fps_with_inference, avg_inference_time = process_video(video_path, with_inference=True)
    print(f"Average FPS with inference: {avg_fps_with_inference:.2f}")
    print(f"Average Inference Time per Frame: {avg_inference_time:.4f}s")

if __name__ == "__main__":
    main()

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @isuchy, thank you for your detailed report and for using Ultralytics 🚀! 

We recommend referring to the [Ultralytics Docs](https://docs.ultralytics.com) for guidance, where you can explore extensive [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) examples. Many commonly asked questions may already be addressed there. 

Since this seems to be a 🐛 Bug Report, if you haven’t already, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) (MRE) to help us identify and address the issue effectively. Your detailed code snippet and logs are immensely helpful—thank you for including those! If you observe any deviations or have a smaller example that isolates the issue further, please share it.

## Quick Suggestions

To address possible issues, ensure the following:
1. You are using the latest `ultralytics` package and dependencies. Please upgrade your environment with:
    ```bash
    pip install -U ultralytics
    ```
   This will ensure your issue is not due to an outdated version. For compatibility, ensure you are using [Python>=3.8](https://www.python.org/) and [PyTorch>=1.8](https://pytorch.org/get-started/locally/).
   
2. Although TFLite is not optimally supported on Jetson devices (as you pointed out), if you are doing a benchmark, you may try experimenting with the NVIDIA TensorRT export instead. It is more suited for Jetson devices and often significantly improves inference speed. Refer to our Export Guides in the [Docs](https://docs.ultralytics.com/modes/export/) for detailed instructions on exporting to TensorRT.

## Join the Community 🌐

- Need real-time advice? Join our [Discord](https://discord.com/invite/ultralytics) 🎧
- Prefer a forum for detailed discussions? Visit [Discourse](https://community.ultralytics.com)
- Share and explore community solutions on [Subreddit](https://reddit.com/r/Ultralytics)

## Environments 🌍

Run YOLO in multiple verified environments with all dependencies preinstalled:  
- Notebooks with free GPU:  
  <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Google Cloud** Deep Learning VM: See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI: See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- Docker Image: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status ✅

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>  

Verify the current build status. Green indicates that all [CI Tests](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) (including exports and inference validations) are passing across macOS, Windows, and Ubuntu.

This is an automated response. One of our team members will review and provide further assistance soon. Let us know if we can assist with anything else in the meantime 😊.

**Comment by glenn-jocher** [comment]
TFLite is not optimized for NVIDIA Jetson platforms, as it lacks GPU support on aarch64 and relies on CPU-only inference, which explains the high latency. For Jetson Xavier NX, consider using TensorRT, which offers far better performance through GPU acceleration. You can export your YOLO model to TensorRT directly using the `model.export(format="engine")` method. Refer to the [DeepStream on NVIDIA Jetson guide](https://docs.ultralytics.com/guides/deepstream-nvidia-jetson/) for detailed setup and configuration steps.

**Comment by Y-T-G** [comment]
This sounds more like a TFLite bug than Ultralytics.

You can try using the standalone example [here](https://github.com/ultralytics/ultralytics/tree/main/examples/YOLOv8-TFLite-Python). If the inference times still high, then it's not an Ultralytics issue.

**Comment by isuchy** [comment]
@glenn-jocher,
let me say again that I am just doing benchmarks of various models on different platforms and I am not intending to deploy this solution into production.
Yes, it's pretty clear that TF Lite is lacking GPU support for aarch64 and inferencing runs on the CPU, but what's surprising is the horrible performance degradation. For example, on a Raspberry Pi 5, the same model takes on average "only" 5.63s for float32 and 1.35s for int8 quantized weights.

@Y-T-G thanks,  thanks for the hint, I will try to install tensorflow-gpu, but wheels in PYPI are usually not built for aarch64, so it probably won't work, though I will try. For JetPack5.1 there is the latest [build of TF](https://docs.nvidia.com/deeplearning/frameworks/install-tf-jetson-platform/index.html) in version 2.12, which is quite old. 
What concerns me is also the fact that the inference of the quantized model takes twice as much time as float32. This could really indicate a bug in the TF.

**Comment by isuchy** [comment]
You are right, it is a matter of hardware implementation and it absolutely depends on the instruction set of the particular CPU. A similar issue has been addressed [here](https://github.com/ARM-software/armnn/issues/511), although it's pretty old, from the days of Raspberry Pi 4 and since then it seems not much has changed.
I also found a pretty overview [page](https://medium.com/@noel.B/yolo-models-on-ncnn-faster-or-slower-a-technical-breakdown-03d36612c921), which also more or less unambiguously confirms that on the arm64 CPU the quantized NCNN is the winner and on the GPU clearly TensorRT. That should bring the dilemma to an end. 
I'm a bit irritated by the claim that TFLite's strength lies in Built-in optimizations for ARM devices, which I've clearly disproved by testing.

> TFLite (TensorFlow Lite):
> 
> Focus: Designed specifically for mobile and embedded devices.
> 
> Strengths: Built-in optimizations for ARM devices, support for quantized models, and integration with TensorFlow workflows.
> 
> Drawbacks: Limited support for non-TensorFlow models compared to NCNN’s broader compatibility.

**Comment by glenn-jocher** [comment]
Thank you for sharing your insights and testing results. You are correct that performance highly depends on hardware-specific optimizations and instruction sets. For ARM-based devices, alternatives like NCNN or TensorRT often yield superior results compared to TFLite, especially when leveraging device-specific acceleration. For Jetson Xavier NX, TensorRT remains the recommended choice for optimal performance, as outlined in our [TensorRT guide](https://docs.ultralytics.com/integrations/tensorrt/). Let us know if we can assist further!
